### PR TITLE
Transform api delete

### PR DESCRIPTION
**Pull request by lobdelle**
*Issue #, if available:*
Delete Transforms API

*Description of changes:*
Added Delete Transforms API, along with unit and Integration tests
Handles single or multiple deletes. Result is idempotent and will not fail if some Transforms do not exist. 
Returns exception if any requested Transforms are still enabled. 

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

### CODE REVIEW

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/DeleteTransformsRequest.kt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Throws(IOException::class)
+    constructor(sin: StreamInput) : this(
+        ids = sin.readStringList()
+    )
+
+    override fun validate(): ActionRequestValidationException? {
+        var validationException: ActionRequestValidationException? = null
+        if (ids.isEmpty()) {
+            validationException = addValidationError("List of ids to delete is empty", validationException)
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/DeleteTransformsRequest.kt:22]: can you add a test case for this, by setting the REST request id to be empty string

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/DeleteTransformsAction.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.amazon.opendistroforelasticsearch.indexmanagement.transform.action.delete
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/DeleteTransformsAction.kt:1]: missing copyright disclosure for all new files

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+) : HandledTransportAction<DeleteTransformsRequest, BulkResponse>(
+    DeleteTransformsAction.NAME, transportService, actionFilters, ::DeleteTransformsRequest
+) {
+
+    override fun doExecute(task: Task, request: DeleteTransformsRequest, actionListener: ActionListener<BulkResponse>) {
+
+        // TODO: if metadata id exists delete the metadata doc else just delete transform
+
+        // Use Multi-Get Request
+        var getRequest = MultiGetRequest()
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt:33]: ``var -> val``

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                    actionListener.onFailure(e)
+                }
+            }
+
+            override fun onFailure(e: Exception) = actionListener.onFailure(e)
+        })
+    }
+
+    private fun bulkDelete(response: MultiGetResponse, ids: List<String>, actionListener: ActionListener<BulkResponse>) {
+        var enabledIDs = mutableListOf<String>()
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt:52]: `var -> val`

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        response.responses.forEach {
+            // TODO: Check if the source is actually transform document
+            if (it.response != null && it.response.isExists) {
+                val source = it.response.source
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt:56]: Can you add a check that the document is indeed transform type before including this in delete? and if the response is failed

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            }
+        }
+
+        if (enabledIDs.isNotEmpty()) {
+            actionListener.onFailure(ElasticsearchStatusException(
+                "[$enabledIDs] transform(s) are enabled, please disable them before deleting them", RestStatus.CONFLICT
+            ))
+        }
+
+        var bulkDeleteRequest = BulkRequest()
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt:71]: `var -> val`

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestDeleteTransformActionIT.kt [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import org.elasticsearch.client.ResponseException
+import org.elasticsearch.rest.RestStatus
+import org.elasticsearch.test.junit.annotations.TestLogging
+
+@TestLogging(value = "level:DEBUG", reason = "Debugging tests")
+@Suppress("UNCHECKED_CAST")
+class RestDeleteTransformActionIT : TransformRestTestCase() {
+    @Throws(Exception::class)
+    fun `test deleting a transform`() {
+
```
Comment (thalurur) [inline src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestDeleteTransformActionIT.kt:17]: minor: remove the empty line

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestDeleteTransformActionIT.kt [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        var transform = randomTransform()
```
Comment (thalurur) [inline src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestDeleteTransformActionIT.kt:18]: ``var -> val``

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestDeleteTransformActionIT.kt [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+import com.amazon.opendistroforelasticsearch.indexmanagement.transform.randomTransform
+import org.elasticsearch.client.ResponseException
+import org.elasticsearch.rest.RestStatus
+import org.elasticsearch.test.junit.annotations.TestLogging
+
+@TestLogging(value = "level:DEBUG", reason = "Debugging tests")
+@Suppress("UNCHECKED_CAST")
+class RestDeleteTransformActionIT : TransformRestTestCase() {
+    @Throws(Exception::class)
+    fun `test deleting a transform`() {
```
Comment (thalurur) [inline src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestDeleteTransformActionIT.kt:16]: Can you add test case with enabled true as well?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    DeleteTransformsAction.NAME, transportService, actionFilters, ::DeleteTransformsRequest
+) {
+
+    override fun doExecute(task: Task, request: DeleteTransformsRequest, actionListener: ActionListener<BulkResponse>) {
+
+        // TODO: if metadata id exists delete the metadata doc else just delete transform
+
+        // Use Multi-Get Request
+        val getRequest = MultiGetRequest()
+        request.ids.forEach { id ->
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt:49]: just to make sure the source is always fetched can you add this?
```
       val includes = arrayOf(
            "${Transform.TRANSFORM_TYPE}.${Transform.ENABLED_FIELD}"
        )
       val fetchSourceContext = FetchSourceContext(true, includes, emptyArray())
       request.ids.forEach { id ->
            getRequest.add(MultiGetRequest.Item(INDEX_MANAGEMENT_INDEX, id).fetchSourceContext(fetchSourceContext))
        }
```

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+            getRequest.add(MultiGetRequest.Item(INDEX_MANAGEMENT_INDEX, id))
+        }
+
+        client.multiGet(getRequest, object : ActionListener<MultiGetResponse> {
+            override fun onResponse(response: MultiGetResponse) {
+                try {
+                    bulkDelete(response, request.ids, actionListener)
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt:56]: Can you check if the response is failed before calling `bulkDelete`, you can sample of the responses and see that its failed

```
                    // response is failed only if managed index is not present
                    if (response.responses.first().isFailed) {
                        actionListener.onFailure(
                            ElasticsearchStatusException(
                                "Cluster missing system index $INDEX_MANAGEMENT_INDEX, cannot execute the request", RestStatus.BAD_REQUEST
                            )
                        )
                        return
                    }
```

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+            actionListener.onFailure(ElasticsearchStatusException(
+                "[$sourceMissingIDs] are missing their source documents!", RestStatus.NOT_FOUND
+            ))
+        }
+
+        if (enabledIDs.isNotEmpty()) {
+            actionListener.onFailure(ElasticsearchStatusException(
+                "[$enabledIDs] transform(s) are enabled, please disable them before deleting them", RestStatus.CONFLICT
+            ))
+        }
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt:106]: I think this can be cleaned up a little - code, errorMessage, missing return statments

```
        val enabledIDs = mutableListOf<String>()
        val notTransform = mutableListOf<String>()

        response.responses.forEach {
            if (it.response.isExists) {
                val source = it.response.source
                val enabled = (source["transform"] as Map<*, *>?)?.get("enabled") as Boolean?
                if (enabled == null) {
                    notTransform.add(it.id)
                }
                if (enabled == true) {
                    enabledIDs.add(it.id)
                }
            }
        }

        if (notTransform.isNotEmpty()) {
            actionListener.onFailure(ElasticsearchStatusException(
                "$notTransform IDs are not transforms!", RestStatus.BAD_REQUEST
            ))
            return
        }

        if (enabledIDs.isNotEmpty()) {
            actionListener.onFailure(ElasticsearchStatusException(
                "$enabledIDs transform(s) are enabled, please disable them before deleting them", RestStatus.CONFLICT
            ))
            return
        }
```

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+        val bulkDeleteRequest = BulkRequest()
+        for (id in ids) {
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/action/delete/TransportDeleteTransformsAction.kt:109]: missing setting the refresh policy 

```
bulkDeleteRequest.refreshPolicy = WriteRequest.RefreshPolicy.IMMEDIATE
```

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestDeleteTransformActionIT.kt [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+            val deleteResponse = client().makeRequest("DELETE",
+                "$TRANSFORM_BASE_URI/${transform.id}")
+            fail("Expected an Exception")
+        } catch (e: Exception) {
+            assertEquals("Expected ElasticsearchStatusException", ResponseException::class, e::class)
+        }
+    }
+
+    @Throws(Exception::class)
+    fun `test deleting a transform that doesn't exist in exiting config index`() {
```
Comment (thalurur) [inline src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/resthandler/RestDeleteTransformActionIT.kt:61]: Minor: Exiting -> existing

**Review by thalurur — APPROVED** [review]
